### PR TITLE
report version ranges as part of audit

### PR DESCRIPTION
**Pull request by CJ-Wright-Citadel**

### CODE REVIEW

File: symbol_exporter/api_match_audit.py [hunk 1]
```diff
@@ -12,10 +14,11 @@
...
 from tqdm import tqdm
 
-from symbol_exporter.api_match import find_supplying_version_set
+from symbol_exporter.api_match import extract_artifacts_from_deps, find_supplying_version_set
 from symbol_exporter.ast_db_populator import make_json_friendly
 from symbol_exporter.ast_symbol_extractor import version, builtin_symbols
 from symbol_exporter.db_access_model import WebDB
+from symbol_exporter.tools import channel_list, find_version_ranges
 
 audit_version = "2.3"
```
Comment (CJ-Wright) [inline symbol_exporter/api_match_audit.py:23]:
```suggestion
audit_version = "2.4"
```

**Comment by CJ-Wright** [comment]
@chrisburr Can you take a look at this? I think this is the next step toward generating a repodata patch.

I am making this comment solely in my personal capacity and am not conveying any rights to any intellectual property of any third parties.

File: tests/test_tools.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+from symbol_exporter.tools import find_version_ranges
+
+
+def test_all_versions_acceptable():
+    assert find_version_ranges(["1.0", "1.1", "1.1.1"], ["1.0", "1.1", "1.1.1"]) == ">=1.0,<=1.1.1"
```
Comment (chrisburr) [inline tests/test_tools.py:5]: Does this want to fail closed or fail open? i.e. when `1.2` is published should it be incorrectly marked as incompatible or should it risk being marked as compatible?

For now I'd be inclined to fail open and make the upper limit be skipped or default to semver, i.e. `>=1.0` or `">=1.0,<2.0"`.

File: symbol_exporter/api_match.py [hunk 3]
```diff
@@ -62,3 +62,20 @@ def find_supplying_version_set(volume, get_symbol_table_func=web_interface.get_s
             else:
                 supplies[top_level_symbol] &= set(artifacts)
     return supplies or set(), bad_symbols
+
+
+def extract_artifacts_from_deps(deps):
+    # TODO: places where multiple artifacts from different packages
+    # satisfy (eg backport pkgs, google-requests error, etc)
+    versions_by_package = {}
```
[the remark this answers is not in the document]
Comment (CJ-Wright-Citadel) [reply symbol_exporter/api_match.py:70]: We're setting this with a bit more nuance I think, so I'll leave this one by the side for the time being.

File: symbol_exporter/api_match.py [hunk 4]
```diff
@@ -62,3 +62,20 @@ def find_supplying_version_set(volume, get_symbol_table_func=web_interface.get_s
...
+    for import_name, artifacts in deps.items():
+        version_set = set()
+        for artifact_path in artifacts:
+            package_name, channel, arch, artifact = artifact_path.split("/")
+            artifact_name, version, build_string = artifact.split("-")
+            version_set.add(version)
+        if package_name not in versions_by_package:
+            versions_by_package[package_name] = version_set
+        else:
+            versions_by_package[package_name] = version_set & versions_by_package[package_name]
```
[the remark this answers is not in the document]
Comment (CJ-Wright-Citadel) [reply symbol_exporter/api_match.py:80]: See above

**Pull request merged**